### DISCUSSION TITLE
Getting Stack Overflow error when using chumsky

### DISCUSSION THREAD

**Discussion by FussballAndy** [zesterer/chumsky#58] [Q&A] [answered]
So I tried implementing chumsky as a parser into a lang I'm working on, however when executing the test I wrote, I get a stack overflow error.

Posting the entire code here would be a huge thing so, I put it into a repo. It can be found in this branch: https://github.com/FussballAndy/unreal-lang/tree/chumsky

The files with chumsky content are the following:
- [`ulc_parser/chumsky_parser.rs`](https://github.com/FussballAndy/unreal-lang/blob/chumsky/crates/ulc_parser/src/chumsky_parser.rs)
- [`ulc_parser/chumsky_parser/`](https://github.com/FussballAndy/unreal-lang/tree/chumsky/crates/ulc_parser/src/chumsky_parser)
- [`ulc_parser/chumsky`](https://github.com/FussballAndy/unreal-lang/blob/chumsky/crates/ulc_parser/tests/chumsky.rs) (overflowing test)

And files with AST are located in [`ulc_ast`](https://github.com/FussballAndy/unreal-lang/tree/chumsky/crates/ulc_ast/src)

I mainly used the nano rust example for reference, if this information is helpful to anybody.

In order to only build and run the necessary test you can run the following `cargo test -p ulc_parser --test chumsky`.

I hope for anyone trying to help me, it is not too bad that the code is embedded in an entire project instead of some code snippets. Additionally, I hope the entire structure of me having expressions as well as statements being completely separated is not too big of a deal too. 

It would be very nice if someone could help me here hence so far I really like chumsky as a parser.

If there happen to be any questions about my code don't hesitate to ask them!

Note: I put this up as a discussion as I am not sure if it is my code being not well written or a problem with the api.

**Comment by zesterer** [comment, marked as answer]
I've not fully looked into the code, but this sounds very typical of left-recursion. Chumsky is, underneath, a PEG parser and left recursion will cause infinite recursion.

You can resolve this by restating your grammar in a way that avoids left recursion.

For example, the left-recursive pattern

```
expr = (expr + "+" + atom) | atom
```

can be restated as

```
expr = atom + ("+" + atom)*
```

[This section](https://github.com/zesterer/chumsky/blob/master/tutorial.md#parsing-binary-operators) in the tutorial about binary operators demonstrates using this technique to properly parse binary operators with precedence in this way.

**Reply by FussballAndy** [reply]
Thank you very much, After a little bit of debugging I got it working! 🎉
